### PR TITLE
Protect api

### PR DESCRIPTION
**Pull request by aguspina**

### CODE REVIEW

File: app/controllers/api/students_controller.rb [hunk 1]
```diff
@@ -14,7 +16,15 @@ def user_params
...
-      @course = Course.find_by(slug: Mumukit::Auth::Slug.join_s(params.to_unsafe_h))
+      @course = Course.find_by(slug: @slug)
+    end
+
+    def set_slug
+      @slug = Mumukit::Auth::Slug.join_s params.to_unsafe_h
+    end
+
+    def protect!
+      @api_client.protect! :janitor, @slug
```
Comment (flbulgarelli) [inline app/controllers/api/students_controller.rb]: Be aware that there are some routes that don't require janitor permissions

File: spec/controllers/courses_api_controller_spec.rb [hunk 2]
```diff
@@ -1,6 +1,8 @@
 require 'rails_helper'
 
 describe Api::CoursesController, type: :controller do
+  before { @request.env["HTTP_AUTHORIZATION"] = api_client.token }
```
Comment (flbulgarelli) [inline spec/controllers/courses_api_controller_spec.rb]: Move this line to a helper

File: config/initializers/active_record_base.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class ActiveRecord::Base
+  def self.update_or_create!(attributes)
+    obj = first || new
```
Comment (flbulgarelli) [inline config/initializers/active_record_base.rb]: You can use `tap` here

File: app/models/user.rb [hunk 4]
```diff
@@ -5,26 +5,38 @@ class User < ApplicationRecord
...
-    self.uid = email
+    self.uid ||= email
   end
 
   def notify!
     NotificationMode.notify_event! 'UserChanged', user: self.as_json
   end
 
+  def set_permissions!
+    Mumukit::Auth::Store.new(:permissions).tap do |db|
```
Comment (flbulgarelli) [inline app/models/user.rb]: Use `Store#set!`

File: app/models/course.rb [hunk 5]
```diff
@@ -11,7 +11,7 @@ class Course < ApplicationRecord
   belongs_to :organization
 
   def add_student!(user)
-    student = User.where(uid: user[:uid]).first_or_create(user)
+    student = User.where(uid: user[:email]).first_or_create(user)
```
Comment (flbulgarelli) [inline app/models/course.rb]: :warning: uid != email

**Comment by aguspina** [comment]
@flbulgarelli please review again!

File: app/controllers/api/students_controller.rb [hunk 6]
```diff
@@ -1,10 +1,12 @@
 module Api
   class StudentsController < BaseController
+    before_action :set_slug, only: :create
     before_action :set_course, only: :create
+    before_action :protect!, only: :create
 
     def create
-      @course.add_student!(user_params)
-      render json: { status: :created }
+      user = @course.add_student(user_params)
```
Comment (flbulgarelli) [inline app/controllers/api/students_controller.rb]: Why have you removed the `!` ?

File: app/controllers/api/users_controller.rb [hunk 7]
```diff
@@ -20,5 +21,9 @@ def user_params
     def set_user
       @user = User.find_by(uid: params[:id])
     end
+
+    def protect_all!
+      @api_client.protect_all! user_params[:permissions]
```
Comment (flbulgarelli) [inline app/controllers/api/users_controller.rb]: I don't like the `protect_all` method name. It is not expressive enough - it doesn't reflect that we are protecting a _delegation_.

File: app/models/concerns/with_permissions.rb [hunk 8]
```diff
@@ -9,5 +9,23 @@ module WithPermissions
     def self.parse!(params)
       params.merge!(permissions: Mumukit::Auth::Permissions.load(params[:permissions]))
     end
+
+    def add_student_permission!(grant)
+      add_permission! :student, grant
+    end
+
+    def update_permissions!(new_permissions)
+      self.update! permissions: permissions.merge(Mumukit::Auth::Permissions.parse(new_permissions))
```
Comment (flbulgarelli) [inline app/models/concerns/with_permissions.rb]: `self` can be avoided here

File: config/initializers/permissions.rb [hunk 9]
```diff
@@ -0,0 +1,6 @@
+class Mumukit::Auth::Permissions
+  def protect_all!(other)
+    other ||= {}
```
Comment (flbulgarelli) [inline config/initializers/permissions.rb]: Why are you accepting a nil input?

File: config/initializers/active_record_base.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+class ActiveRecord::Base
+  def self.update_or_create(attributes)
```
Comment (flbulgarelli) [inline config/initializers/active_record_base.rb]: What about  `assign_first`?

File: config/routes.rb [hunk 11]
```diff
@@ -2,7 +2,7 @@
   root to: 'organizations#index'
   resources :organizations, only: [:index, :show, :edit, :create, :update]
   resources :users, only: [:index, :show, :edit, :create, :update]
-  resources :api_clients, only: [:new, :create]
+  #resources :api_clients, only: [:new, :create]
```
Comment (flbulgarelli) [inline config/routes.rb]: ?

**Pull request merged**